scripts/train_lm.py

```diff
@@ -146,12 +146,39 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
 
 def FormatMetaparameters(metaparameters):
     out = []
+    # get the maximum length of decimals of metaparameters
+    decimal_length = 0
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
-        out.append(x)
-
+        temp_length = len(str(param))
+        if decimal_length < temp_length:
+            decimal_length = temp_length
+    decimal_length -= 2
+    
+    temp_original = metaparameters[0]
+    temp = round(temp_original, 3)
+    for param in metaparameters[1:]:
+        round_index = 3
+        x_original = param
+        x = round(x_original, round_index)
+        while(temp == x and round_index <= decimal_length):
```

Comment:
It is not really hard to use that, I think it has at three advantages that I can see:
i) We can manipulate the way of rounding, such like ‘ROUND_DOWN’ or ‘ROUND_UP’
ii) 0.00000123123 will be 0.00000123 if we set the precision to 3 (i.e. this is what you mention as ‘AppendDigitUtilNonZero’)
iii) something like str(0.000001) gives ‘1e-6’, which can be bad. Decimal gives a good str representation of it, like ‘0.000001’

However, there are few issue that Dan mention
i) what kind of collision is permissible? My understanding is collision in  "data-type weights” is not OK, otherwise it is fine?
ii) Should we check validity in this function, or pass it to other function to check validity
iii) If we find something suspicious, my understanding is to warn user instead of terminating everything?

If you interested in using this code, I will fix these issues so you may use it if you like.


On Aug 18, 2016, at 10:30 PM, Wang Jian <notifications@github.com> wrote:

> In scripts/train_lm.py:
> 
> > -        if x == '0.00':
> > -            x = '{:.3g}'.format(param)
> > -        out.append(x)
> > -
> > +        temp_length = len(str(param))
> > +        if decimal_length < temp_length:
> > +            decimal_length = temp_length
> > +    decimal_length -= 2
> > +    
> > +    temp_original = metaparameters[0]
> > +    temp = round(temp_original, 3)
> > +    for param in metaparameters[1:]:
> > +        round_index = 3
> > +        x_original = param
> > +        x = round(x_original, round_index)
> > +        while(temp == x and round_index <= decimal_length):
> Zhouyang, your code looks plausible, although I don't know how to use the decimal module.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>